cli/popper/commands/cmd_add.py

```diff
@@ -16,55 +16,61 @@
     short_help='Add a pipeline from popperized repositories on github.'
 )
 @click.argument('pipeline', required=True)
+@click.option(
+    '--folder',
+    help='Folder where to store the new pipeline, relative to project root.',
+    show_default=True,
+    default='pipelines'
+)
 @pass_context
-def cli(ctx, pipeline):
+def cli(ctx, pipeline, folder):
     """Add a pipeline to your repository from the existing popperized
     repositories on github. The pipeline argument is provided as owner/repo/
-    pipeline. For example, popper add popperized/quiho-popper/single-node
-    adds the single-node pipeline from the quiho-popper repository.
+    pipeline. For example, 'popper add popperized/quiho-popper/single-node'
+    adds the 'single-node' pipeline from the 'quiho-popper' repository from the
+    'popperized' organization.
     """
-    try:
-        owner, repo, pipeline_name = pipeline.split('/')
-    except ValueError:
-        pu.fail("See popper add --help for more info.")
+    if len(pipeline.split('/')) != 3:
+        pu.fail("Bad pipeline name. See 'popper add --help' for more info.")
+
+    owner, repo, pipe_name = pipeline.split('/')
+
+    config = pu.read_config()
+
+    if pipe_name in config['pipelines']:
+        pu.fail("Pipeline {} already in repo.".format(pipe_name))
 
     project_root = pu.get_project_root()
+    pipelines_dir = os.path.join(project_root, folder)
+
+    if not os.path.exists(pipelines_dir):
+        os.mkdir(pipelines_dir)
 
-    pipelines_dir = os.path.join(project_root, 'pipelines')
-    if os.path.exists(pipelines_dir):
-        pass
-    else:
-        os.chdir(project_root)
-        os.mkdir('pipelines')
-
-    pipeline_path = os.path.join(pipelines_dir, pipeline_name)
-    os.chdir(pipelines_dir)
-    os.mkdir(pipeline_name)
-    os.chdir(pipeline_path)
-
-    pipeline_url = 'https://github.com/{}/{}/tree/master/pipelines/{}' \
-        .format(owner, repo, pipeline_name)
-    download_url = 'https://github-download-only-a-folder.glitch.me/dl?url={}'\
-        .format(pipeline_url)
-
-    pu.info("Downloading pipeline {} ... ".format(pipeline_name))
-    r = requests.get(download_url)
-    if r.status_code == 200:
-        try:
-            z = zipfile.ZipFile(BytesIO(r.content))
-            z.extractall()
-        except BadZipfile:
-            z.close()
-    else:
+    gh_url = 'https://github.com/{}/{}/archive/master.zip'.format(owner, repo)
+
+    pu.info("Downloading pipeline {}... ".format(pipe_name))
+
+    r = requests.get(gh_url)
+    if r.status_code != 200:
         pu.fail("Unable to fetch the pipeline. Please check if the name" +
                 " of the pipeline is correct and the internet is connected")
+    z = zipfile.ZipFile(BytesIO(r.content))
```

Comment:
We might need to open the zipfile with a `with` clause to close the file automatically. I mean something like this :- 

```
with zipfile.ZipFile(BytesIO(r.content)) as z :
    z.extractall()
```

this would help us avoid any unwanted errors due the file not being closed properly. 

Also maybe a try and catch block for the BadZipfile errors should be kept since it seems to be a standard practice when using zip files. What do you think?

Reply:
I added the `with`. We might want to open a separate issue for properly handling exceptions all over our codebase (I think we have more places where we could catch known exceptions), so that when an error occurs, more user friendly messages are displayed